instantout/reservation/store.go

```diff
@@ -0,0 +1,300 @@
+package reservation
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"time"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightninglabs/loop/fsm"
+	"github.com/lightninglabs/loop/loopdb"
+	"github.com/lightninglabs/loop/loopdb/sqlc"
+	"github.com/lightningnetwork/lnd/keychain"
+)
+
+// BaseDB is the interface that contains all the queries generated
+// by sqlc for the reservation table.
+type BaseDB interface {
+	// CreateReservation stores the reservation in the database.
+	CreateReservation(ctx context.Context,
+		arg sqlc.CreateReservationParams) error
+
+	// GetReservation retrieves the reservation from the database.
+	GetReservation(ctx context.Context,
+		reservationID []byte) (sqlc.Reservation, error)
+
+	// GetReservationUpdates fetches all updates for a reservation.
+	GetReservationUpdates(ctx context.Context,
+		reservationID []byte) ([]sqlc.ReservationUpdate, error)
+
+	// GetReservations lists all existing reservations the client has ever
+	// made.
+	GetReservations(ctx context.Context) ([]sqlc.Reservation, error)
+
+	// UpdateReservation inserts a new reservation update.
+	UpdateReservation(ctx context.Context,
+		arg sqlc.UpdateReservationParams) error
+
+	// ExecTx allows for executing a function in the context of a database
+	// transaction.
+	ExecTx(ctx context.Context, txOptions loopdb.TxOptions,
+		txBody func(*sqlc.Queries) error) error
+}
+
+// SQLStore manages the reservations in the database.
+type SQLStore struct {
+	baseDb BaseDB
+
+	nowTime func() time.Time
+}
+
+// NewSQLStore creates a new SQLStore.
+func NewSQLStore(db BaseDB) *SQLStore {
+	nowTime := time.Now
```

Comment:
nit: a nice thing to have is to inject a clock as a dependency, so we can set  deterministic expectations in tests.